### PR TITLE
feat(gli/utils.py): add class HjsonTemplate

### PR DESCRIPTION
**Pull request by wood-ghost**

### CODE REVIEW

**Review by xingjian-zhang — CHANGES_REQUESTED** [review]
Generally looks good to me👍, I will leave more comments once I try the new functionality on my machine. For now, I have a few concerns:
- Please make sure the code can pass all (style) tests including pydocstyle/pylint/pycodestyle. You can run these tests locally by `make pystyle` and `make pylint`
- Also write a notebook to demo the usage of the new class. For example, on cora dataset.

File: gli/utils.py [hunk 1]
```diff
@@ -392,3 +392,366 @@ def save_data(prefix, **kwargs):
     for key, matrix in sparse_arrays.items():
         sp.save_npz(f"{prefix}_{key}.sparse.npz", matrix)
         print("Save sparse matrix", key, "to", f"{prefix}_{key}.sparse.npz")
+
+class HjsonTemplate(object):
```
[the remark this answers is not in the document]
Comment (jiaqima) [reply gli/utils.py:396]: Agreed.

**Comment by jiaqima** [comment]
Please also fill in some descriptions in each section of the PR template, especially "Description" and "How Has This Been Tested?"

File: gli/utils.py [hunk 2]
```diff
@@ -392,3 +392,366 @@ def save_data(prefix, **kwargs):
...
+            d.update({"key" : k})
+        return d
+    
+    def metadata(
+        self,
+        node=dict(),
+        edge=dict(),
+        graph=dict(),
+        citation=None,
+        is_heterogeneous=True
```
Comment (xingjian-zhang) [inline gli/utils.py:420]: We might want to make the interface even easier for users. For example, for homogeneous graph, we may want the function to take in these parameters:

```python
def graph(self,
    edges,
    node_list,
    edge_list=None,
    node_attrs={},
    edge_attrs={},
    graph_attrs={},
    citation=None,
)
```
The first two inputs are corresponding to our required properties. The `edge_list` is an optional property. And all other inputs are optional.

Comment (xingjian-zhang) [reply gli/utils.py:420]: Also, should we split the saving interface of homo- and hetero-graph. What do
you think? @jiaqima

File: gli/utils.py [hunk 3]
```diff
@@ -392,3 +392,366 @@ def save_data(prefix, **kwargs):
...
+            if "key" not in graph.get("_EdgeList"):
+                self._missing_features_messages_warning.append("%s missing 'key' in '_EdgeList'" % self._name)
+                graph.update({"key" : "edge_list"})
+        self._graph = graph 
+        self._citation = citation
+        self._is_heterogeneous = is_heterogeneous
+
+        # return missing_features_messages_error, missing_features_messages_warning
+    
+    def save_metadata(self):
```
Comment (xingjian-zhang) [inline gli/utils.py:452]: Perhaps we can directly save the data when user calls the `metadata` function?
What prevents us from doing so?

File: gli/utils.py [hunk 4]
```diff
@@ -392,3 +392,366 @@ def save_data(prefix, **kwargs):
     for key, matrix in sparse_arrays.items():
         sp.save_npz(f"{prefix}_{key}.sparse.npz", matrix)
         print("Save sparse matrix", key, "to", f"{prefix}_{key}.sparse.npz")
+
+class HjsonTemplate(object):
+    def __init__(
+        self,
+        name
```
Comment (xingjian-zhang) [inline gli/utils.py:399]: We can add a `path` param here to specify the storage path.

File: gli/utils.py [hunk 5]
```diff
@@ -392,3 +392,366 @@ def save_data(prefix, **kwargs):
...
+                test_set.update({"file" : "%s_task.npz" % self._name})
+            if "key" not in test_set:
+                missing_features_messages_warning.append("%s missing 'key' in 'test_set'" % self._name)
+                test_set.update({"key" : "val"})
+        self._test_set = test_set
+
+        if ((self._num_samples != 0) or (self._train_ratio != 0) or (self._val_ratio != 0) or (self._test_ratio!= 0)) and (self._train_set or self._val_set or self._test_set):
+            missing_features_messages_error.append("%s has conflict in splits in %s" % (self._name, self._type))
+
+        return missing_features_messages_error, missing_features_messages_warning
```
Comment (xingjian-zhang) [inline gli/utils.py:516]: Perhaps we can directly print the error at the end of the function and raise a
`ValueError` then.

**Comment by xingjian-zhang** [comment]
I also suggest to separate the storage of metadata and task. Contributors might want to save multiple tasks, and it is unclear how to do so in current implementation. The better solution might be one class `MetadataConfiguration` and one class `TaskConfiguration` or something like that.

File: gli/utils.py (hunk 2 above)
Comment (jiaqima) [reply gli/utils.py:420]: Yeah, I think it's better to split the interface for these two, which may help significantly simplify the API

Comment (xingjian-zhang) [reply gli/utils.py:420]: Add a class called `Attribute` with member `description`, `data`, `type`, `format`.

**Pull request closed without merging**